### DISCUSSION TITLE
test harness infrastructure question

### DISCUSSION THREAD

**Discussion by pprindeville** [libuv/libuv#3418] [Q&A]
I've got a PR that's in progress, but I wanted to add some test coverage to it (as I don't have access to all the platforms that `libuv` runs on to do exhaustive testing... and I'd rather add tests that would then add some degree of regression protection in the future).  Problem is I'm not sure how, and I've not seen an example in `test/` that does something similar.

I want to run (depending on the platform) various commands, capture the output, parse it, and then use that as the source of truth to compare against the values that my code being tested returns.

What's the appropriate way to do this?  And what support is baked into the test harness for doing so?

**Comment by bnoordhuis** [comment]
There's no out of the box support in the test harness but you can use uv_spawn() to spawn a subprocess and capture its output.

**Reply by pprindeville** [reply]
Is there a good example of running an external command-line utility that I can use as a template?

**Reply by bnoordhuis** [reply]
The uv_spawn() tests in test/test-spawn.c are a good starting point.

**Reply by pprindeville** [reply]
I'd want to post-process the output via regex to extract the interesting fields... this is starting to snowball.

For instance:

```
% ip -4 -o addr show
1: lo    inet 127.0.0.1/8 scope host lo\       valid_lft forever preferred_lft forever
2: ens3    inet 192.168.4.33/24 brd 192.168.4.255 scope global dynamic noprefixroute ens3\       valid_lft 41118sec preferred_lft 41118sec
```

I'd want to extract the interface name (`lo`, `ens3`) and then compare the associated fields (ipv4 address and prefix `127.0.0.1/8` and `192.168.4.33/24` respectively, or broadcast address `192.168.4.255` where present) and compare those against what's been discovered by `libuv`.

**Reply by bnoordhuis** [reply]
Regular expressions are unfortunately out of the question (no good cross-platform support and third-party libraries aren't something we want to take on as dependencies) but you can use `fscanf()` or `sscanf()` to do field-based parsing.

That said, you're probably in for a rough time making it work on multiple platforms. `ip` is a Linux-ism; on the BSDs, you have to shell out to `ifconfig` and deal with any platform differences in its output.

**Reply by pprindeville** [reply]
> Regular expressions are unfortunately out of the question (no good cross-platform support and third-party libraries aren't something we want to take on as dependencies) but you can use `fscanf()` or `sscanf()` to do field-based parsing.
> 
> That said, you're probably in for a rough time making it work on multiple platforms. `ip` is a Linux-ism; on the BSDs, you have to shell out to `ifconfig` and deal with any platform differences in its output.

I didn't want to put it into `libuv`, obviously... just into the `test/` harness.  For Linux/UNIX it shouldn't be a problem, but for Windows we'd need `std::regex` wrappers to call from C.

**Reply by bnoordhuis** [reply]
I understand but that's precisely what I mean. For various reasons we're unlikely to accept patches that mix C and C++.

On the subject of `regcomp()` and friends: once you start using that, you also have to start thinking about `setlocale()` and when to call or not call it because the locale affects pattern matching. It tends to snowball in complexity.

**Reply by pprindeville** [reply]
> I understand but that's precisely what I mean. For various reasons we're unlikely to accept patches that mix C and C++.
> 
> On the subject of `regcomp()` and friends: once you start using that, you also have to start thinking about `setlocale()` and when to call or not call it because the locale affects pattern matching. It tends to snowball in complexity.

I imagine for output from commands, you'd want the command to run in the `C` locale, and parse its output similarly.  What about `libpcre2`?  That's a single C-based library that runs on everything, and is POSIX-compliant.

**Reply by bnoordhuis** [reply]
I expect there is going to be very little enthusiasm in taking on libpcre2 as a dependency. One of libuv's selling features is the simplicity with which you can build and use it, you only need libc.

There are some single header file regular expression libraries out there that might be acceptable as a middle ground, but caveat lector, they also need to be license compatible.
